### PR TITLE
Firefox: Pass navigation tests

### PR DESCRIPTION
**Pull request by Meir017**
related to #111

### CODE REVIEW

**Comment by Meir017** [comment]
@kblok could the failures be due to certificate issues?

**Comment by kblok** [comment]
@Meir017 yes. It must be that. Could you try installing certificates on the windows build?

File: src/PlaywrightSharp.Tests/TestConstants.cs [hunk 1]
```diff
@@ -1,93 +1,95 @@
...
+using Microsoft.Extensions.Logging.Xunit;
+using PlaywrightSharp.Chromium;
+using PlaywrightSharp.Firefox;
+using Xunit.Abstractions;
+
+namespace PlaywrightSharp.Tests
+{
+    internal static class TestConstants
+    {
+        public const string BROWSER_CDN = "https://playwright2.blob.core.windows.net";
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/TestConstants.cs:16]: `BrowserCDN`

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs [hunk 2]
```diff
@@ -11,12 +11,29 @@ public class PlaywrightSharpLoader
...
         internal static SimpleServer Server { get; private set; }
         internal static SimpleServer HttpsServer { get; private set; }
+        internal static string ContentRoot { get; private set; }
 
         internal static async Task SetupAsync()
         {
-            var downloaderTask = TestConstants.GetNewBrowserType().CreateBrowserFetcher().DownloadAsync();
-            Server = SimpleServer.Create(TestConstants.Port, TestUtils.FindParentDirectory("PlaywrightSharp.TestServer"));
-            HttpsServer = SimpleServer.CreateHttps(TestConstants.HttpsPort, TestUtils.FindParentDirectory("PlaywrightSharp.TestServer"));
+            string host = !TestConstants.IsChromium ? TestConstants.BROWSER_CDN : null;
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs:18]: Why do we need this?

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs [hunk 3]
```diff
@@ -11,12 +11,29 @@ public class PlaywrightSharpLoader
...
+            {
+                if (percentage != e.ProgressPercentage)
+                {
+                    percentage = e.ProgressPercentage;
+                    Console.WriteLine($"[{TestConstants.Product}] downloading browser {percentage}%");
+                }
+            };
+            Console.WriteLine($"Downloading browser from {host}...");
+            var downloaderTask = browserFetcher.DownloadAsync();
+            ContentRoot = TestUtils.FindParentDirectory("PlaywrightSharp.TestServer");
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs:34]: Why do we need this?

File: src/PlaywrightSharp.Tests/Page/GoBackTests.cs [hunk 4]
```diff
@@ -25,11 +27,11 @@ public async Task ShouldWork()
 
             var response = await Page.GoBackAsync();
             Assert.True(response.Ok);
-            Assert.Equal(TestConstants.EmptyPage, response.Url);
+            Assert.Contains(TestConstants.EmptyPage, response.Url);
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Page/GoBackTests.cs:30]: Why contains?

File: src/PlaywrightSharp.Tests/Page/GoToTests.cs [hunk 5]
```diff
@@ -78,7 +82,7 @@ public async Task ShouldCaptureIframeNavigationRequest()
             Assert.Equal(TestConstants.ServerUrl + "/frames/one-frame.html", response.Url);
 
             Assert.Equal(2, Page.Frames.Length);
-            Assert.Same(Page.FirstChildFrame(), requestFrame);
+            Assert.Same(Page.Frames[1], requestFrame);
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Page/GoToTests.cs:85]: Why are we changing this?

File: src/PlaywrightSharp.Tests/TestConstants.cs [hunk 6]
```diff
@@ -1,93 +1,95 @@
-using System;
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/TestConstants.cs:1]: ouch

File: src/PlaywrightSharp/Request.cs [hunk 7]
```diff
@@ -149,7 +149,11 @@ internal void SetResponse(Response response)
         {
             Response = response;
             _waitForResponseTsc.TrySetResult(response);
-            WaitForFinished.ContinueWith(_ => _waitForFinishedTsc.TrySetResult(response), TaskScheduler.Default);
+            response.Finished.ContinueWith(
+                _ =>
+            {
+                return _waitForFinishedTsc.TrySetResult(response);
+            }, TaskScheduler.Default);
```
Comment (kblok) [inline src/PlaywrightSharp/Request.cs:156]: I would give one more indent to this. And TaskScheduler.Default in a new line

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs (hunk 2 above)
Comment (Meir017) [reply src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs]: to download the patched version.
see https://github.com/kblok/playwright/blob/b8bb58e48b47200a6d4311b48a67132945609d21/browser_patches/tools/check_cdn.sh

File: src/PlaywrightSharp.Tests/Page/GoToTests.cs (hunk 5 above)
Comment (Meir017) [reply src/PlaywrightSharp.Tests/Page/GoToTests.cs:85]: this is the correct way, how playwright is doing it

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs (hunk 3 above)
Comment (Meir017) [reply src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs:34]: for `GoToTests.ShouldOverrideReferrerPolicy`

File: src/PlaywrightSharp.Tests/Page/GoBackTests.cs (hunk 4 above)
Comment (Meir017) [reply src/PlaywrightSharp.Tests/Page/GoBackTests.cs:30]: this is the correct assert based on playwright

File: src/PlaywrightSharp.Tests/Page/GoToTests.cs [hunk 8]
```diff
@@ -533,15 +549,18 @@ public async Task ShouldOverrideReferrerPolicy()
             Server.SetRoute("/grid.html", context =>
             {
                 context.Response.Headers["Referrer-Policy"] = "no-referrer";
-                return Task.CompletedTask;
+                string file = File.ReadAllText(Path.Combine(PlaywrightSharpLoader.ContentRoot, "wwwroot", "grid.html"));
```
Comment (Meir017) [inline src/PlaywrightSharp.Tests/Page/GoToTests.cs:552]: not sure about a clean way for this

File: src/PlaywrightSharp.Tests/Page/GoToTests.cs (hunk 5 above)
Comment (kblok) [reply src/PlaywrightSharp.Tests/Page/GoToTests.cs:85]: We brought this recipe from Puppeteer. In .NET the order of the frames is not guaranteed. Browser callbacks are processed in parallel an the `ConcurrentDictionary` does not guarantee an order.

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs (hunk 2 above)
Comment (kblok) [reply src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs]: What would be the difference between that and this https://github.com/hardkoded/playwright-sharp/blob/master/src/PlaywrightSharp.Firefox/FirefoxBrowserType.cs#L273 ?

File: src/PlaywrightSharp.Tests/Page/GoToTests.cs (hunk 8 above)
Comment (kblok) [reply src/PlaywrightSharp.Tests/Page/GoToTests.cs:552]: I implemented a Subscribe method here https://github.com/hardkoded/playwright-sharp/blob/a958f565cf46db5b3e7739df27970ed132844884/src/PlaywrightSharp.Tests/Webkit/ProvisionalTests.cs#L32

You add the `Referrer-Policy` on a `Subscribe` and then let the server handle the file.

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs (hunk 3 above)
Comment (kblok) [reply src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs:34]: See my comment on the test.

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs (hunk 2 above)
Comment (Meir017) [reply src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs]: not sure. they seem to be different but might contain the same archives
https://playwright2.blob.core.windows.net and https://playwright.azureedge.net

Comment (kblok) [reply src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs]: I don't see where check_cdn.sh is being used in tests.

File: src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs [hunk 9]
```diff
@@ -14,7 +14,23 @@ public class PlaywrightSharpLoader
 
         internal static async Task SetupAsync()
         {
-            var downloaderTask = TestConstants.GetNewBrowserType().CreateBrowserFetcher().DownloadAsync();
+            string host = !TestConstants.IsChromium ? TestConstants.BrowserCDN : null;
+            var browserFetcher = TestConstants.GetNewBrowserType().CreateBrowserFetcher(new BrowserFetcherOptions
+            {
+                Host = host
+            });
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/BaseTests/PlaywrightSharpLoader.cs:21]: I don't think we need this.

File: src/PlaywrightSharp.Tests/TestConstants.cs [hunk 10]
```diff
@@ -13,6 +13,7 @@ namespace PlaywrightSharp.Tests
 {
     internal static class TestConstants
     {
+        public const string BrowserCDN = "https://playwright2.blob.core.windows.net";
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/TestConstants.cs:16]: Nor this.

File: src/PlaywrightSharp.Tests/Page/GoBackTests.cs [hunk 11]
```diff
@@ -25,11 +27,11 @@ public async Task ShouldWork()
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Page/GoBackTests.cs:27]: GoBackTests is not working for me locally.

**Comment by kblok** [comment]
We are all green here @Meir017. Do you want to take a final look?

**Pull request merged**